src/sections/upload-dataset-files/UploadDatasetFiles.tsx

```diff
@@ -43,19 +44,24 @@ export const UploadDatasetFiles = ({ fileRepository: fileRepository }: UploadDat
 
   const fileUploadFinished = (file: File) => {
     const key = FileUploadTools.key(file)
-    setUploadFinished((x) => x.add(key))
     setUploadingToCancelMap((x) => {
       x.delete(key)
       return x
     })
     releaseSemaphore(file)
   }
 
+  const canUpload = (file: File) =>
+    !uploadingToCancelMap.has(FileUploadTools.key(file)) &&
+    !FileUploadTools.get(file, fileUploaderState).failed &&
+    !FileUploadTools.get(file, fileUploaderState).done
+
   const uploadOneFile = (file: File) => {
     const key = FileUploadTools.key(file)
-    if (uploadingToCancelMap.has(key) || uploadFinished.has(key)) {
+    // sanity check: should not happen
+    /*if (!canUpload(file)) {
```

Comment:
Please remove this commented code to not bring confusion if its not needed.

Reply:
I removed it. It was a nice defensive-programming sanity check, but I could not keep it without it interfering with the coverage testing.

Reply:
Well, if you think is a good sanity check but is too hard to test that rare case I think is better to leave it ( but working not commented) and add in that case an istanbul ignore.